fix(Pilot Sheet): pilot imports ungrouped, immediately visible

msprijatelj opened the pull request:
# Description
This PR refines #1647 such that newly-imported pilots ~~import their group with them, and~~ are immediately visible. This PR also makes pilot sorting/reindexing more robust.  Finally, this PR changes the default display state for new groups to "visible", ~~and adds more detail to the Import Pilot message to include Pilot Group~~.

EDIT: Updated such that ~~newly-created and~~ imported pilots do NOT import their group, and are instead Ungrouped.  Removed Pilot Group from Import Pilot message.  Keeping the sorting/reindexing and the default display states, as I think they meaningfully add to the project.

EDIT 2: Moved the "ungrouping" to pilot imports only, so that cloned pilots retain the group of their original pilot (a feature that I feel will be more convenient than always sending a "new" pilot to Ungrouped).  Completely new pilots lack a group in the first place, so this changes nothing for them.

## Issue Number
Closes #1650

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

msprijatelj commented on src/features/pilot_management/Roster/components/add_panels/ImportDialog.vue:
```diff
@@ -23,10 +23,12 @@
           //
           <b>{{ pilot.Name }}</b>
+          to group 
+          <b>{{ pilot.Group ? pilot.Group : 'Ungrouped' }}</b>
```
Add a little more detail about the group the Pilot will import into.

msprijatelj commented on src/features/pilot_management/Roster/index.vue:
```diff
@@ -171,7 +171,7 @@ export default Vue.extend({
     newGroupMenu: false,
     tempGroups: [],
-    shown: [],
+    hidden: [],
```
Flipping from `shown` to `hidden` so that newly-imported pilots & their groups are immediately visible, without needing to push the pilot to "shown".

msprijatelj commented on src/features/pilot_management/store/index.ts:
```diff
@@ -10,6 +10,21 @@ async function savePilots(pilots: Pilot[]) {
+function reindexPilots(pilots: Pilot[]) {
+  pilots.forEach((p, i) => p.SortIndex = i)
+}
+
```
Move some commonly-used helper functions outside of the Mutations/Actions.

msprijatelj commented on src/features/pilot_management/store/index.ts:
```diff
@@ -44,20 +59,16 @@ export class PilotManagementStore extends VuexModule {
 
   @Mutation
   private [ADD_PILOT](payload: Pilot): void {
+    payload.SortIndex = this.Pilots.length
```
This index isn't permanent; it's simply to position the pilot at the "end" of the `Pilots` array for sorting purposes.

msprijatelj commented on src/features/pilot_management/store/index.ts:
```diff
@@ -104,12 +118,17 @@ export class PilotManagementStore extends VuexModule {
+    this.Pilots.forEach((p: Pilot, i) => {
+      if (p.Group === payload) {
+        p.Group = ''
+        p.SortIndex = this.Pilots.length + i
```
This sort index is qualitative; as long as relative order is preserved, the sort/reindex below will clean up the group deletion properly, and deposit the newly-groupless pilots in their original order at the end of the "Ungrouped" group.

msprijatelj commented on src/features/pilot_management/store/index.ts:
```diff
@@ -165,6 +184,7 @@ export class PilotManagementStore extends VuexModule {
 
   @Action
   public addPilot(payload: { pilot: Pilot; update: boolean }): void {
+    this.context.commit(ADD_GROUP, payload.pilot.Group)
```
This is the major piece that will fix #1650: Simply making sure the group is appended to PilotGroups prior to adding the Pilot themselves (and subsequently sorting/reindexing as appropriate).

msprijatelj commented:
Once bitten, twice shy, so please feel free to QA this change before approval/merging.  I believe I covered my bases for displaying imported pilots immediately, but verification is always appreciated.

TranslucentSabre commented:
I personally don't think it is very user-friendly to import the group along with the pilot from the vault.
As a GM, I will want to group my players pilots together regardless of which group each of them have selected for their own characters.

msprijatelj commented:
> 
> 
> I personally don't think it is very user-friendly to import the group along with the pilot from the vault.
> As a GM, I will want to group my players pilots together regardless of which group each of them have selected for their own characters.

You know what, I completely agree.  Removed the Pilot Group import by setting a new pilot's group to `''` and fixed the import message to match.  Keeping the rest of the changes for now; I think they'll make the code more maintainable.
